### PR TITLE
odin infra (api) stack

### PR DESCRIPTION
**Pull request by joakimmoller**
VPC, mongo, cluster and service in one separate stack.

All other odin-smr related services should use the VPC to connect to the mongodb and the postgres-server

Odin-api will be deployed as an docker image and will on deploy reload the service (from a boto3 script).

### CODE REVIEW

File: .github/workflows/cd.yaml [hunk 1]
```diff
@@ -0,0 +1,44 @@
+name: Continuous Deployment
+
+on:
+  push:
+    branches:
+      - main
```
Comment (skymandr) [inline .github/workflows/cd.yaml:6]: Do we want deploy on every push to master?

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      uses: actions/setup-python@v2
+      with:
+        python-version: "3.10"
+
+    - name: Python install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install tox
+
+    - name: Set AWS credentials
```
Comment (skymandr) [inline .github/workflows/ci.yaml]: Do we need credentials for tests?

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+[tool.black]
+line-length = 88
```
Comment (skymandr) [inline pyproject.toml:2]: Didn't we want defaults?

File: tests/unit/test_odin_infrastructure_stack.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import aws_cdk
+import aws_cdk.assertions as assertions
+
+from odin_infrastructure.odin_api_stack import OdinAPIStack
+from odin_infrastructure.config import ODIN_AWS_ACCOUNT, ODIN_AWS_REGION
+
+
+# example tests. To run these tests, uncomment this file along with the example
+# resource in odin_infrastructure/odin_infrastructure_stack.py
```
Comment (skymandr) [inline tests/unit/test_odin_infrastructure_stack.py]:
```suggestion
```

File: odin_infrastructure/odin_api_stack.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        nat_gateway_provider = ec2.NatProvider.instance(
+            instance_type=ec2.InstanceType("t3.small")
+        )
+        vpc: ec2.IVpc = ec2.Vpc(
+            self,
+            "OdinVPC",
+            nat_gateway_provider=nat_gateway_provider,
+            max_azs=2,
+            nat_gateways=1,
+            vpc_name="OdinVPC",
```
Comment (skymandr) [inline odin_infrastructure/odin_api_stack.py:24]: Is this name all other services need to know, for their stacks to set up permissions etc?

File: odin_infrastructure/odin_cluster.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            log_group_name="/Odin/OdinApi",
+            removal_policy=RemovalPolicy.DESTROY,
+            retention=aws_logs.RetentionDays.SIX_MONTHS,
+        )
+
+        logging = aws_ecs.AwsLogDriver(stream_prefix="OdinAPI", log_group=log_group)
+
+        psql_bucket = aws_s3.Bucket.from_bucket_name(
+            scope, "OdinPsqlBucket", "odin-psql"
+        )
```
Comment (skymandr) [inline odin_infrastructure/odin_cluster.py:36]: What is this bucket for?

File: odin_infrastructure/mongo.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            scope,
+            "OdinMongoInstanceRole",
+            assumed_by=iam.ServicePrincipal("ec2.amazonaws.com"),
+            managed_policies=[
+                iam.ManagedPolicy.from_aws_managed_policy_name(
+                    "CloudWatchLogsFullAccess"
+                ),
+            ],
+        )
+        mongodb_repo = textwrap.dedent(
```
Comment (skymandr) [inline odin_infrastructure/mongo.py:39]: Nice trick with `textwrap`!

**Review by skymandr — APPROVED** [review]
A few small questions and comments, but looks splendid over all.

File: odin_infrastructure/odin_api_stack.py (hunk 5 above)
Comment (joakimmoller) [reply odin_infrastructure/odin_api_stack.py:24]: Yes, unless that service need access to a bucket or so.

File: .github/workflows/cd.yaml (hunk 1 above)
Comment (joakimmoller) [reply .github/workflows/cd.yaml:6]: yes!

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (joakimmoller) [reply .github/workflows/ci.yaml]: maybe not...

File: pyproject.toml (hunk 3 above)
Comment (joakimmoller) [reply pyproject.toml:2]: default is 80 +10% - so this ensures default :)

**Review by Baapity — APPROVED** [review]
bump-approve

**Pull request merged**